libraries/AP_Notify/LP5562.cpp

```diff
@@ -0,0 +1,118 @@
+/*
+  LP5562 I2C driver
+*/
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* LED driver for LP5562 */
+
+#include "LP5562.h"
+
+#if AP_NOTIFY_LP5562_ENABLED
+
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define LP5562_LED_BRIGHT  255    // full brightness
+#define LP5562_LED_MEDIUM  170    // medium brightness
+#define LP5562_LED_DIM     85     // dim
+#define LP5562_LED_OFF     0      // off
+
+enum class Register {
+    ENABLE  = 0x00,
+    B_PWM   = 0x02,
+    G_PWM   = 0x03,
+    R_PWM   = 0x04,
+    CONFIG  = 0x08,
+    LED_MAP = 0x70,
+};
+
+LP5562::LP5562(uint8_t bus, uint8_t addr)
+    : RGBLed(LP5562_LED_OFF, LP5562_LED_BRIGHT, LP5562_LED_MEDIUM, LP5562_LED_DIM)
+    , _bus(bus)
+    , _addr(addr)
+{
+}
+
+bool LP5562::init(void)
+{
+    _dev = std::move(hal.i2c_mgr->get_device(_bus, _addr));
+    if (!_dev) {
+        return false;
+    }
+    WITH_SEMAPHORE(_dev->get_semaphore());
+
+    _dev->set_retries(10);
+
+    //  values here are taken literally from 7.3.2 in the datasheet.
+    //  See the simulator for register breakdown.
+
+    // chip enable:
+    _dev->write_register((uint8_t)Register::ENABLE, 0b1000000);
+
+    // start-up-delay:
+    hal.scheduler->delay_microseconds(500);
+
+    // use internal clock:
+    _dev->write_register((uint8_t)Register::CONFIG, 0b00000001);
+
+    // set direct PWM control:
+    _dev->write_register((uint8_t)Register::LED_MAP, 0b00000000);
+
+    _dev->set_retries(1);
+
+    _dev->register_periodic_callback(20000, FUNCTOR_BIND_MEMBER(&LP5562::_timer, void));
```

Comment:
we need to check the write_register calls, and not install the callback unless they succeed
even better would be a detection that this really is a LP5562, perhaps do a RESET then read LED_MAP and check for 0x71?